proposals/0026/README.md

```diff
@@ -0,0 +1,275 @@
+---
+id: 0026
+title: Support for the Kubernetes Resource Model
+status: ideation
+authors: Micah Hausler <mhausler@amazon.com>
+---
+
+## Summary
+
+This is a proposal to architect Tinkerbell as an application using the [Kubernetes Resource Model][krm].
+It is a rearchitecture of the 'control plane' backend, and leaves the 'data plane' components of Tink workers and actions unchanged.
+
+[krm]: https://github.com/kubernetes/community/blob/master/contributors/design-proposals/architecture/resource-management.md
+
+## Goals and not Goals
+
+Goals:
+* **Compatibility with existing Tink workers, workflow execution, and actions.**
+* **More easily support non-request serving controllers in Tinkerbell.**
+  In this architecture, controllers like PBnJ could leverage Kubernetes primitives like [Custom Resource Definitions][crds] (CRDs), [WATCH APIs][watch], and [Field Management][fm] to complete workflow steps.
+* **Reduce the security surface of the Tinkerbell API.**
+  Tinkerbell is a high-value component of data center infrastructure, so protection of DHCP infrastructure, BMC/IPMI management, needs to be treated as such.
+  Implementing multiple authorization modes is a non-trivial task.
+  The fewer APIs, and authorization options, and lines of code that exist, the fewer opportunities there are for security issues to arise.
+* **Support a highly-available architecture.**
+  Postgres is a fantastic database, but managing high-availability with graceful failover is not trivial.
+  Using an alternative data store that better supports failure would better help operators to have higher availability and not require downtime for upgrades or failover.
+* **Migrate existing components**.
+  Hegel, Boots, and Tinkerbell API will be modified to use Kubernetes as the datastore
+* **Support migration of existing installations.**
+  Migration tooling and documentation will be provided for existing Workflows, Hardware, and Templates stored in Tinkerbell.
+  Migration should be as straightforward as:
+  * Creating Tinkerbell CRDs in the Kubernetes API
+  * Running a provided migration command
+  * Restarting each Tinkerbell component targeting Kubernetes (ex: `--k8s-mode=true`)
```

Comment:
> This argument supposes that the k8s data backend support will be optional. This asserts that another non-goal would be to remove the existing API interface.

 I'll descope PG deprecation for another proposal.

> This argument also supposes that we will not be dissecting the existing service runtimes from their frontends and data access models. That's a change that I think could be related to this work so it could be enumerated among non-goals. If we were to do so, we might have a hegel-pg and hegel-k8s binary, for example, or hegel-backend-pg and hegel-backend-k8s plugins. But it would appear that today, at least, removing PostgreSQL support is a non-goal.
> I would love to see future proposals for both of these changes :-)

I'll try to articulate this better. The main thrust here is to have two modes of operating Tinkerbell: 
* One as it exists today (architecturally and existing APIs)
* One with the proposed architecture and reduced Tinkerbell API surface. This second mode would be enabled on components (Tink API, boots, hegel) with a feature flag

Reply:
>>  This argument supposes that the k8s data backend support will be optional. This asserts that another non-goal would be to remove the existing API interface.

>  ~I'll descope PG deprecation for another proposal.~

I'm going back on this and keeping it in the proposal. I think there are 4 separate stages in this transition.
1. Support full existing API and Postgres backend (Today)
2. Support full existing API and Postgres backend, as well as K8s resource model (KRM)
3. Deprecate most of the existing API and Postgres backend. Continue to build out and support KRM  
4. Remove existing non-worker APIs, only support KRM.

Do we need to define how long stages 2 and 3 last?